add google-oauthlib-tool

proppy opened the pull request:
Commandline tool to generate credentials using 3LO oauth2 flow.

See:
https://github.com/GoogleCloudPlatform/google-auth-library-python/issues/152

/cc @vchudnov-g

jonparrott commented:
@vchudnov-g how do you feel about this? My thoughts are that your work will make this code simpler, but that it's still useful in the meantime.

It also seems this overlaps a bit with oauth2l, but I have a *lot* of issues with oauth2l.

proppy commented:
PTAL

jonparrott commented:
This code LGTM, but ISTM that since this is now `googlesamples-oauthlib-tool` it might make sense for it to live in the `googlesamples` namespace (`googlesamples.authtool.__main__.py` for example).

Such would necessitate a separate library, and if that's a path you wanna go down @proppy, I'm happy to help with setup/CI/etc.

proppy commented:
> Such would necessitate a separate library

Can't we have the google-auth-oauthlib ship packages under `google` and `googlesamples` ?

jonparrott commented:
> Can't we have the google-auth-oauthlib ship packages under google and googlesamples ?

It actually *intentionally* doesn't ship packages under `google` right now. The reason is that eventually this library will merge upstream with `google-auth` once it's ready. When that happens, it'll still be possible to install both `google-auth` (which will provide `google.oauth2.flow` etc) and `google-auth-oauthlib` (which provides `google_auth_oauthlib.flow`) without them stomping on each other. This means that users can upgrade without breaking their code, as installing and using the standalone library will continue to work. We'll just put a deprecation warning telling folks to use the upstream version but we won't break anyone.

proppy commented:
oh ok, maybe it's best to rename the executable to `google-oauthlib-tool` for now and move it to `googlesamples` once we move the library to `google`.

Thoughts?

proppy commented:
PTAL :)

jonparrott commented on .coveragerc:
```diff
@@ -4,6 +4,7 @@ branch = True
 [report]
 omit =
     */conftest.py
+    */__main__.py
```
Can you write some tests for main instead of ignoring it in coverage?

proppy replied:
Done. Splitted the bulk of __main__ in __init__.

proppy commented on google_auth_oauthlib/tool/__main__.py:
```diff
@@ -0,0 +1,118 @@
+    return flow.credentials
+
+
+def credentials_to_dict(credentials):
```
should we use the same format as https://github.com/GoogleCloudPlatform/google-auth-library-python/blob/0c09c734ba8ed24f2c9f3ea8a1a190f29aa679b6/google/auth/_cloud_sdk.py#L83 ?

proppy commented:
PTAL, now there is proper test for `__main__`

jonparrott commented on google_auth_oauthlib/tool/__main__.py:
```diff
@@ -0,0 +1,78 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Command line tool for fetching credentials using 3LO OAuth2 flow.
```
Recommendation: I've very much been concerned with preciseness of language in this library, maybe:

```
Command-line tool for obtaining authorization and credentials from a user.

This tool uses the OAuth 2.0 Authorization Code grant as described in `section 1.3.1 of RFC 6749`_
and implemeted by :class:`google_auth_oauthlib.flow.Flow`.

..  _section 1.3.1 of RFC 6749: https://tools.ietf.org/html/rfc6749#section-1.3.1
```

I'd recommend applying the terminology to the help message as well.

proppy replied:
Done.

proppy commented:
PTAL

- updated docstring
- reduced the footprint of the tool package (everything is in `__main__` and tested)
- removed `access_token` from the serialized credentials

jonparrott commented on google_auth_oauthlib/tool/__main__.py:
```diff
@@ -0,0 +1,108 @@
+        config_path = os.path.dirname(credentials_file)
+        if not os.path.isdir(config_path):
+            os.makedirs(config_path)
+        with open(credentials_file, 'w') as f:  # pylint: disable=invalid-name
```
use `io.open` over `open`.

proppy replied:
Done

jonparrott commented on google_auth_oauthlib/tool/__main__.py:
```diff
@@ -0,0 +1,108 @@
+
+
+if __name__ == '__main__':
+    main()  # pylint: disable=no-value-for-parameter
```
Will you add a comment as to why you disabled this? e.g. `# pylint doesn't realize that click has changed the function signature.`

proppy replied:
Done

jonparrott commented on google_auth_oauthlib/tool/__main__.py:
```diff
@@ -0,0 +1,108 @@
+:class:`google_auth_oauthlib.flow.Flow`.
+
+This tool is intended to bootstrap development in situation when the
+application cannot easily run the 3LO OAuth2 flow, i.e: in an embedded
```
super nit: 3LO is never defined in the rfc, so please let's stick to terminology used in the rfc. Maybe:

> This tool is intended for assist developers in obtaining credentials for testing applications where it may not be possible or easy to run a complete OAuth 2.0 authorization flow, especially in the case of code samples or embedded devices without input / display capabilities.

proppy replied:
Done

jonparrott commented on google_auth_oauthlib/tool/__main__.py:
```diff
@@ -0,0 +1,108 @@
+This is not intended for production where the main application should
+run the 3LO OAuth2 flow to get authorization from the users.
+
+..  __: https://tools.ietf.org/html/rfc6749#section-1.3.1
```
Nit: please use rst named references:

```python
.. _section 1.3.1 of RFC 6749: https://tools.ietf.org/html/rfc6749#section-1.3.1
```

Just in case we add more links.

proppy replied:
Done
